### PR TITLE
[Core][State Observability] Get API using List endpoints + filtering on ids

### LINKED ISSUE #25716: [Core][State Observability API] Implement Get APIs.
### Description

As described in REP, we should implement `get` APIs for state API. `get` API will return a single entity's state.

https://github.com/ray-project/enhancements/pull/8/files

## Implementation
In the first version, we can simply return the value from `ray list [resources] --filter id [ID]`. 

In the mid-term (after alpha), we'd like to query more than 1 source to obtain detailed data, which includes;

Actor: Currently running tasks, task history (func_name -> count & stats mapping), scheduling details, queue size.
Node: Machine metrics, top 20 high process memory usage. 
Task: Scheduling state in more details

Additionally, we'd like to optimize the performance by filtering the ID from the source side instead of from the API server. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This is a first implementation of GET APIs for
- nodes
- actors
- placement groups
- workers
- tasks
- objects

E.g.

```
# CLI
(dev) ➜  ray git:(ricky/obs-get) ray get nodes cab26304d105caa6f2100908f7b461ef9ed244984ec30b4b46f953f9
---
node_id: cab26304d105caa6f2100908f7b461ef9ed244984ec30b4b46f953f9
node_ip: 172.31.47.143
node_name: 172.31.47.143
resources_total:
    CPU: 8.0
    memory: 16700517582.0
    node:172.31.47.143: 1.0
    object_store_memory: 8350258790.0
state: ALIVE


# Python 
from ray.experimental.state.api import get_node
from ray.experimental.state.common import NodeState

node :NodeState = get_node(<id>)
print(node)

```

We currently do not support getting specific resources by id for 'jobs' and 'runtime-envs'
- jobs: it is not exposing id to be queried easily yet 
- runtime envs: it doesn't have an id associated. 

### TODO:
- it uses list endpoints + filtering as for now, future iterations will implement GET-specific endpoints and interaction with raylet/GCS with point query APIs. 
- Unit testing for `state_manager` for GET endpoints when implemented. 
- Getting jobs by id

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [X] Integration tests
     - `pytest python/ray/tests/test_state_api.py`
   - [X] Manual CLI tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/api.py [hunk 1]
```diff
@@ -146,6 +146,208 @@ def list(
...
+        ]
+    ]:
+        """Get resources states by id
+
+        Args:
+            resource_name: Resource names, i.e. 'workers', 'actors', 'nodes',
+                'placement_groups', 'tasks', 'objects'.
+                'jobs' and 'runtime-envs' are not supported yet.
+            id: ID for the resource, i.e. 'node_id' for nodes.
+            options: List options. See `GetApiOptions` for details.
```
Comment (rkooo567) [inline python/ray/experimental/state/api.py:172]:
```suggestion
            options: Get options. See `GetApiOptions` for details.
```

File: python/ray/experimental/state/api.py [hunk 2]
```diff
@@ -146,6 +146,208 @@ def list(
...
+        options: Optional[GetApiOptions],
+        _explain: bool = False,
+    ) -> Optional[
+        Union[
+            ActorState,
+            PlacementGroupState,
+            NodeState,
+            WorkerState,
+            TaskState,
+            ObjectState,
```
Comment (rkooo567) [inline python/ray/experimental/state/api.py:162]: It should be Union[ObectState, List[ObjectState]]?

File: python/ray/experimental/state/api.py [hunk 3]
```diff
@@ -146,6 +146,208 @@ def list(
...
+            StateResource.ACTORS: "actor_id",
+            StateResource.PLACEMENT_GROUPS: "placement_group_id",
+            StateResource.WORKERS: "worker_id",
+            StateResource.TASKS: "task_id",
+            StateResource.OBJECTS: "object_id",
+        }
+        if resource not in RESOURCE_ID_KEY_NAME:
+            raise ValueError(f"Can't get {resource.name} by id.")
+
+        params["filter_keys"] = [RESOURCE_ID_KEY_NAME[resource]]
```
Comment (rkooo567) [inline python/ray/experimental/state/api.py:209]: Why don't we just call self.list here?

**Review by rkooo567 — APPROVED** [review]
LGTM. Just a few comments! Let's have the follow-up (actual get endpoints) next month!

File: python/ray/experimental/state/api.py (hunk 3 above)
Comment (rickyyx) [reply python/ray/experimental/state/api.py:209]: We could - was prepping for the separation of get from list later. A bit of pre-mature factoring here.

File: python/ray/experimental/state/api.py (hunk 2 above)
Comment (rickyyx) [reply python/ray/experimental/state/api.py:162]: I see - was thinking of returning a list even if there is only a single `ObjectState` entry. But `Union` will work.

**Comment by rickyyx** [comment]
To resolve conflict and address commens.

**Comment by rkooo567** [comment]
Please remove the author-action-required label when it is ready to merge!

**Comment by rickyyx** [comment]
Ready to merge when tests passed.

**Pull request merged**